text/0000-discriminant-bits.md

```diff
@@ -0,0 +1,120 @@
+- Feature Name: discriminant_bits
+- Start Date: 2019-04-01
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add methods to `std::mem::Discriminant` which inform of the space necessary for a bitwise representation of an enum's discriminant and the bits itself in an opaque fashion.
+
+# Motivation
+[motivation]: #motivation
+
+Rust encourages using enums to encode data with multiple variants.
+An example of this can be found in the [game of life tutorial][game-of-life-tutorial].
+
+```rust
+enum Cell {
+    Dead  = 0,
+    Alive = 1
+}
+```
+
+Using these enums in collections is wasteful, as each instance reserves at least 1 byte of space.
+Similarly, `std::mem::size_of<Discriminant<Cell>>()` is at least 1 byte.
+For that reason, the book later goes on and replaces `Vec<Cell>` by [`fixedbitset`][game-of-life-exercise], ending up with a much less intuitive implementation.
+
+If it were possible to read the exact necessary size and the bit representation the descriminant, we could have a `PackedBits<T>` that uses exactly as much space as necessary.
+
+This allows for an efficient representation of discriminant sets, which is both useful for simple enums, but also for crating an index of all discriminant values present in collection.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Disciminant data
+
+`Discriminant::bit_size` is a method to retrieve the minimal number in bits necessary to represent this discriminant.
+
+```rust
+const fn bit_size() -> usize { }
+```
+
+This number is not subject to optimisation, so e.g. `Option<&str>` reports a bitsize of `1`.
+
+For example:
+
+```rust
+enum Cell {
+    Dead = 0,
+    Alive = 1,
+}
+
+enum RGB {
+    Red,
+    Green,
+    Blue
+}
+
+Discriminant<Cell>::bit_size() == 1
+Discriminant<Option<&str>>::bit_size() == 1
+Discriminant<RGB>::bit_size() == 2
+```
+
+This information can be used to pack multiple discriminants easily for efficient storage and easy indexing.
+
+`Discriminat<T>` gains the methods `into_bits` and `from_bits`:
+
+```rust
+fn into_bits(&self) -> u128
```

Comment:
When I first saw this, I had to wonder what crazy kind of code would have an enum with more than `2^64` variants.  Maybe `#[repr(u128)]` should be mentioned as justification?

Reply:
In my mind it's not just about having `> 2^64` variants. You can have much fewer and use `VariantN = DiscrimExprN` to use up a whole lot of bits.

Reply:
Yes, that was my point.  I had to think for a while before I remembered that explicit discriminants can be assigned.

Reply:
@eddyb had actually recommended using that size, as that would be the final width of the internal `Discriminant` value.